### PR TITLE
Support per-behavior axes

### LINKED ISSUE #39: Support per-behavior axes
Docker compose environment would look like:

``` yaml
services:
    crossdock:
        environment:
            - CROSSDOCK_WAIT_FOR=yarpc-node,yarpc-go,yarpc-python,yarpc-java

            - CROSSDOCK_AXIS_CLIENT=yarpc-node,yarpc-go,yarpc-python,yarpc-java
            - CROSSDOCK_AXIS_SERVER=yarpc-node,yarpc-go,yarpc-python,yarpc-java
            - CROSSDOCK_AXIS_BAD_HOST=yarpc-node

            - CROSSDOCK_BEHAVIOR_ECHO=client,server,transport
            - CROSSDOCK_BEHAVIOR_RAW=client,server
            - CROSSDOCK_BEHAVIOR_JSON=client,server
            - CROSSDOCK_BEHAVIOR_THRIFT=client,server
            - CROSSDOCK_BEHAVIOR_BAD_INBOUND=bad-host,server
            - CROSSDOCK_BEHAVIOR_BAD_OUTBOUND=client,bad-host
```

Note that this separates WAIT_FOR and CLIENT into semantically separate headers (with the same value). The BEHAVIORS header disappears and is implied by the enumeration of behavior environment variables.

### PR DESCRIPTION
**Pull request by jc-fireball**
Resolves #39 
- Create new version of docker-compose.yaml
- Implement new crossdock config parsing logic

@yarpc/yarpc

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,9 +23,11 @@ crossdock:
         - alpha
         - omega
     environment:
-        - CROSSDOCK_CLIENTS=alpha,omega
-        - CROSSDOCK_AXIS_BEHAVIOR=dance,run
+        - CROSSDOCK_WAIT_FOR=alpha,omega
```
Comment (HelloGrayson) [inline README.md:26]: put new lines between `WAIT_FOR`, `AXIS_*`, and `BEHAVIOR_*`

File: docker-compose.yml [hunk 2]
```diff
@@ -4,8 +4,12 @@ crossdock:
         - alpha
         - omega
     environment:
-        - CROSSDOCK_CLIENTS=alpha,omega
-        - CROSSDOCK_AXIS_BEHAVIOR=dance,sing
+      - CROSSDOCK_WAIT_FOR=alpha,omega
```
Comment (HelloGrayson) [inline docker-compose.yml:7]: same here

File: plan/entities.go [hunk 3]
```diff
@@ -15,6 +16,13 @@ type Axis struct {
 	Values []string
 }
 
+// Behavior represents the test behavior will be triggered by crossdock
+type Behavior struct {
+	Name    string
+	Clients string
```
Comment (HelloGrayson) [inline plan/entities.go:25]: `Client`*

File: plan/config.go [hunk 4]
```diff
@@ -12,41 +14,73 @@ const defaultCallTimeout = 5
...
+	fmt.Println(os.Getenv(waitKey))
+	waiters := trimCollection(strings.Split(os.Getenv(waitKey), ","))
+	axes := make(map[string]Axis)
+	behaviors := make(map[string]Behavior)
 	for _, e := range os.Environ() {
-		if !strings.HasPrefix(e, axisKeyPrefix) {
-			continue
-		}
-		d := strings.TrimPrefix(e, axisKeyPrefix)
+		if strings.HasPrefix(e, axisKeyPrefix) {
```
Comment (HelloGrayson) [inline plan/config.go:30]: this if else block is huge - perhaps the contents should be in pure functions?

File: plan/config.go [hunk 5]
```diff
@@ -12,41 +14,73 @@ const defaultCallTimeout = 5
...
-			Name:   key,
-			Values: values,
+			behaviors[key] = behavior
 		}
-		axes = append(axes, axis)
 	}
 	config := &Config{
 		CallTimeout: time.Duration(callTimeout),
-		Clients:     clients,
+		Waiters:     waiters,
```
Comment (HelloGrayson) [inline plan/config.go]: Let's rename this to `WaitForHosts: hosts` - waiters makes me feel like those are the things that will be doing the waiting.

**Comment by HelloGrayson** [comment]
Looks pretty good :+1: 

Let's wait to merge this until we've tested it with YARPC errors branch.

File: plan/config.go [hunk 6]
```diff
@@ -12,41 +14,73 @@ const defaultCallTimeout = 5
...
+			d := strings.TrimPrefix(e, behaviorKeyPrefix)
 
-		pair := strings.SplitN(d, "=", 2)
-		key := strings.ToLower(pair[0])
-		values := strings.Split(pair[1], ",")
-		values = trimCollection(values)
+			pair := strings.SplitN(d, "=", 2)
+			key := strings.ToLower(pair[0])
+			values := strings.Split(pair[1], ",")
+			values = trimCollection(values)
```
Comment (prashantv) [inline plan/config.go]: these 4 lines are exactly the same in both branches. DRY, use a function.

File: plan/config.go [hunk 7]
```diff
@@ -12,41 +14,73 @@ const defaultCallTimeout = 5
...
 	}
+
+	validateConfig(config)
 	return config
 }
 
+func validateConfig(config *Config) {
+	for _, behavior := range config.Behaviors {
+		if _, ok := config.Axes[behavior.Clients]; !ok {
+			log.Fatal("Can't find AXIS environment for: " + behavior.Clients)
```
Comment (prashantv) [inline plan/config.go]: return an error, ideally only the `main` package should use `log.Fatal`.

File: plan/config.go [hunk 8]
```diff
@@ -12,41 +14,73 @@ const defaultCallTimeout = 5
...
+		}
+		for _, param := range behavior.Params {
+			if _, ok := config.Axes[param]; !ok {
+				log.Fatal("Can't find AXIS environment for: " + param)
```
Comment (prashantv) [inline plan/config.go]: same as above

File: plan/new.go [hunk 9]
```diff
@@ -11,15 +13,19 @@ func New(config *Config) *Plan {
 
 func buildTestCases(plan *Plan) []TestCase {
 	var testCases []TestCase
-	for _, client := range plan.Config.Clients {
-		combos := recurseCombinations(plan, client, plan.Config.Axes, make(map[string]string))
-		testCases = append(testCases, combos...)
+	for _, behavior := range plan.Config.Behaviors {
+		for _, client := range plan.Config.Axes[behavior.Clients].Values {
+			fmt.Println(behavior)
```
Comment (prashantv) [inline plan/new.go]: stale print?

**Comment by HelloGrayson** [comment]
@junchaowu will you update the description? It doesn't state the problem or what you accomplished in the PR.

File: plan/new.go [hunk 10]
```diff
@@ -11,15 +13,19 @@ func New(config *Config) *Plan {
...
+			testCases = append(testCases, combos...)
+		}
 	}
 	return testCases
 }
 
-func recurseCombinations(plan *Plan, client string, axes []Axis, args Arguments) []TestCase {
-	if len(axes) == 0 {
+func recurseCombinations(level int, plan *Plan, client string, behavior Behavior, args Arguments) []TestCase {
+	fmt.Println(args, level, len(behavior.Params))
```
Comment (prashantv) [inline plan/new.go]: stale print?

File: plan/new_test.go [hunk 11]
```diff
@@ -3,45 +3,186 @@ package plan
 import (
 	"testing"
 
+	"github.com/bradfitz/slice"
```
Comment (prashantv) [inline plan/new_test.go]: do you really need this dependency? Avoid adding dependencies that save a couple of lines of code. You can very easily create a custom type and use the standard sort package.

File: plan/new_test.go [hunk 12]
```diff
@@ -3,45 +3,186 @@ package plan
...
+				Clients: "client",
+				Params:  []string{"server", "transport"},
+			},
+			"sing": {Name: "sing",
+				Clients: "client",
+				Params:  []string{"server", "transport"},
+			}}})
+
+	slice.Sort(plan.TestCases, func(i, j int) bool {
+		if plan.TestCases[i].Client != plan.TestCases[j].Client {
```
Comment (prashantv) [inline plan/new_test.go:28]: you reference `plan.TestCases[i]` and `plan.TestCases[j]` > 5 times in this function, and it's super long, make a short variable.

``` go
p1 := plan.TestCases[i]
p2 := plan.TestCases[j]

// use p1 and p2
```

File: plan/new_test.go [hunk 13]
```diff
@@ -3,45 +3,186 @@ package plan
...
+			return plan.TestCases[i].Client < plan.TestCases[j].Client
+		}
+		if plan.TestCases[i].Arguments["server"] != plan.TestCases[j].Arguments["server"] {
+			return plan.TestCases[i].Arguments["server"] < plan.TestCases[j].Arguments["server"]
```
Comment (prashantv) [inline plan/new_test.go:32]: you can also DRY up this by doing

``` go
if diff := p1.Arguments["server"] - p2.Arguments["server"]; diff != 0 {
  return diff < 0
}
```

you can do the same for every test you have.

**Comment by prashantv** [comment]
DRY up the code and use smaller functions.

File: plan/config.go [hunk 14]
```diff
@@ -12,41 +14,73 @@ const defaultCallTimeout = 5
...
+		axisKeyPrefix     = "CROSSDOCK_AXIS_"
+		behaviorKeyPrefix = "CROSSDOCK_BEHAVIOR_"
 	)
 	callTimeout, _ := strconv.Atoi(os.Getenv(callTimeoutKey))
 	if callTimeout == 0 {
 		callTimeout = defaultCallTimeout
 	}
-	clients := trimCollection(strings.Split(os.Getenv(clientsKey), ","))
-	var axes []Axis
+	fmt.Println(os.Getenv(waitKey))
```
Comment (abhinav) [inline plan/config.go]: Did you mean to leave this print statement here?

File: plan/config.go (hunk 4 above)
Comment (abhinav) [inline plan/config.go:30]: +1. Move all this repetitive string transformation logic into a separate function with a descriptive name and documentation to give a better idea of what the inputs and outputs are.

File: plan/entities.go (hunk 3 above)
Comment (jc-fireball) [inline plan/entities.go:25]: behavior may have multi clients. so I'd keep using clients

File: plan/new_test.go (hunk 13 above)
Comment (jc-fireball) [inline plan/new_test.go:32]: the value of Arguments is string type so minute operation doesn't work.

**Pull request merged**